### PR TITLE
Ormichae/add action probability ccb response

### LINKED ISSUE #209: Add Action-Probability list as part of ccb response
We cannot change the rankResponse which APS returns to the user without it being considered a breaking change. Currently, our response includes a list of action and a corresponding list of probabilities.  request_multi_slot_decision returns a DecisionResponse which contains the single recommended action along with the corresponding probability. We cannot convert the DecisionResponse into the RankResponse which we return to the user. 

- Add an Action-Probability list to request_multi_slot_decision response for each slot.

### PR DESCRIPTION
**Pull request by ormichae**
[https://github.com/VowpalWabbit/reinforcement_learning/issues/209](url)

Added endpoint for ccb which returns all action probability pairs for each slot.

resolves #209

### CODE REVIEW

**Review by kumpera — CHANGES_REQUESTED** [review]
This PR adds binaries from the output folder.
Could you update the `.gitignore` file to avoid this mistake in the future?

File: include/ranking_response.h [hunk 1]
```diff
@@ -16,41 +16,28 @@
...
-  };
-
   /**
    * @brief choose_rank() returns the action choice using ranking_response.
    * ranking_response contains all the actions and distribution from with the action was sampled.  It also contains the chosen action id and
    * the unique event_id representing the choice.  This unique event_id must be used to report back outcomes against this choice for the online trainer.
    */
   class ranking_response {
+  private:
+	using coll_t = std::vector<action_prob>;
```
Comment (ataymano) [inline include/ranking_response.h:26]: looks like it is cleaner to have dependency directly from slot_ranking::iterator instead of redefining new coll_t type (which is the same as slot_ranking::coll_t, but it is not enforced anyhow)

File: include/ranking_response.h [hunk 2]
```diff
@@ -59,124 +46,119 @@ namespace reinforcement_learning {
...
+	* @brief Move assignment operator for slot detailed.
+	* The underlying data is taken from rvalue reference, and then it is cleared.
+	* @return ranking_response&
+	*/
+	ranking_response& operator=(ranking_response&&) noexcept;
+
+	/**
+	* @brief Copy constructor is removed since implementation will be deleted twice
+	*/
+	ranking_response(const slot_ranking&) = delete;
```
Comment (ataymano) [inline include/ranking_response.h:145]: slot_ranking -> ranking_response?

File: include/ranking_response.h [hunk 3]
```diff
@@ -59,124 +46,119 @@ namespace reinforcement_learning {
...
+
+	/**
+	* @brief assignment operator is removed since implementation will be deleted twice
+	*/
+	ranking_response& operator =(const slot_ranking&) = delete;
```
Comment (ataymano) [inline include/ranking_response.h:150]: slot_ranking -> ranking_response?

File: include/ranking_response.h [hunk 4]
```diff
@@ -59,124 +46,119 @@ namespace reinforcement_learning {
...
-    using iterator = container_iterator<action_prob, coll_t>;
-    using const_iterator = const_container_iterator<action_prob, coll_t>;
-    //! Returns an iterator pointing to the first element of the (action, probability) collection
-    const_iterator begin() const;
-    iterator begin();
-
-    //! Returns an iterator referring to the past-the-end element of the (action, probability) collection.
-    const_iterator end() const;
-    iterator end();
+	  using iterator = container_iterator<action_prob, coll_t>;
```
Comment (ataymano) [inline include/ranking_response.h:154]: probably better just to use slot_ranking::iterator

File: include/ranking_response.h [hunk 5]
```diff
@@ -16,41 +16,28 @@
...
-     * @brief Construct a new ranking response object.
-     *
-     * @param event_id The unique identifier for this interaction.  This event_id must be used when reporting the outcome for this action
-     */
-    ranking_response(char const* event_id);
+	ranking_response() = default;
+	~ranking_response() = default;
+
+	/**
+	* @brief Construct a new reanking response object.
```
Comment (ataymano) [inline include/ranking_response.h:36]: typo

File: include/ranking_response.h [hunk 6]
```diff
@@ -16,41 +16,28 @@
...
+	*
+	* @param join_id The unique identifier for this interaction.  This event_id must be used when reporting the outcome for this action
```
Comment (ataymano) [inline include/ranking_response.h:38]: join_id -> event_id

File: include/ranking_response.h [hunk 7]
```diff
@@ -59,124 +46,119 @@ namespace reinforcement_learning {
...
+	/**
+	* @brief Get the model_id.
+	* Every rank call (single or multi slot) is associated with a unique model used to predict.  A unique model_id
+	* is associated with each unique model. (This is set internally by the API)
+	* @return const char*
+	*/
+	const char * get_model_id() const;
+
+	/**
+	* @brief Clear the slot detailed object so that it can be reused.
```
Comment (ataymano) [inline include/ranking_response.h:124]: looks like all comments here should stay the same - slot_ranking is implementation detail from this class point of view

File: include/ranking_response.h [hunk 8]
```diff
@@ -59,124 +46,119 @@ namespace reinforcement_learning {
...
-
-    /**
-     * @brief Set the model_id.
-     * Every call to choose action is associated with a unique model used to predict.  A unique model_id
-     * is associated with each unique model. (This is set internally by the API).
-     * Input model_id is left in an unspecified but valid state.
-     * @param model_id
-     */
-    void set_model_id(std::string&& model_id);
+	/**
```
Comment (ataymano) [inline include/ranking_response.h:55]: Looks like there are a lot of whitespace changes.
Although github allows to ignore them during PR, they will still make final result inaccurate.
Can you please double check that in your C++ editor's settings you have?:
- TAB = 2 spaces
- replace TAB with spaces

File: include/multi_slot_response_detailed.h [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+
+		int set_slot_at_index(const unsigned int index, slot_ranking&& slot, api_status* status = nullptr);
+
+		void clear();
+		const_iterator_t begin() const;
+		iterator_t begin();
+		const_iterator_t end() const;
+		iterator_t end();
+
+		//what is the no except?? methods 2 and 4 are overlaods for the = operand but what are 1 and 3? what is the &&reference?
```
Comment (olgavrou) [inline include/multi_slot_response_detailed.h:49]: 1 and 3 are move and copy constructors

File: bindings/cs/rl.net.native/rl.net.multi_slot_response_detailed.h [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+	API const char* GetMultiSlotDetailedEventId(reinforcement_learning::multi_slot_response_detailed* slot);
+	API const char* GetMultiSlotDetailedModelId(reinforcement_learning::multi_slot_response_detailed* slot);
+
+	API size_t GetMultiSlotDetailedSize(reinforcement_learning::multi_slot_response_detailed* multi_slot);
+
+	API multi_slot_detailed_enumerator_adapter* CreateMultiSlotDetailedEnumeratorAdapter(reinforcement_learning::multi_slot_response_detailed* multi_slot);
+	API void DeleteMultiSlotDetailedEnumeratorAdapter(multi_slot_detailed_enumerator_adapter* adapter);
+
+	API int MultiSlotDetailedEnumeratorInit(multi_slot_detailed_enumerator_adapter* adapter);
+	API int MultiSlotEnumeratorDetailedMoveNext(multi_slot_detailed_enumerator_adapter* adapter);
```
Comment (ataymano) [inline bindings/cs/rl.net.native/rl.net.multi_slot_response_detailed.h:23]: MultiSlotDetailedEnumeratorMoveNext?

File: include/multi_slot_response_detailed.h [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+		multi_slot_response_detailed(multi_slot_response_detailed&&) noexcept;
+		multi_slot_response_detailed& operator=(multi_slot_response_detailed&&) noexcept;
+		multi_slot_response_detailed(const multi_slot_response_detailed&) = default;
```
Comment (ataymano) [inline include/multi_slot_response_detailed.h:52]: Other similar classes (i.e. ranking_response) have copy constructors/ assignments deleted. Why do we need it here?

Comment (olgavrou) [reply include/multi_slot_response_detailed.h:52]: We are inconsistent, `ranking_response` is as you say, but `multi_slot_response` does the same as shown here. I'd say we should be consistent and follow `ranking_response` for all.

Comment (ataymano) [reply include/multi_slot_response_detailed.h:52]: yes i agree

Comment (ormichae) [reply include/multi_slot_response_detailed.h:52]: That was the source of confusion. I will go ahead and delete the copy constructors from multi_slot_response as well for consistency.

File: include/multi_slot_response_detailed.h (hunk 9 above)
Comment (ormichae) [reply include/multi_slot_response_detailed.h:49]: forgot to remove that comment when I had it answered.

File: bindings/cs/rl.net.native/rl.net.multi_slot_response_detailed.h (hunk 10 above)
Comment (ormichae) [reply bindings/cs/rl.net.native/rl.net.multi_slot_response_detailed.h:23]: messed up the naming on that one. fixed now

File: bindings/cs/rl.net.native/rl.net.live_model.cc [hunk 12]
```diff
@@ -136,7 +136,7 @@ API int LiveModelRequestMultiSlotDecision(livemodel_context_t* context, const ch
   RL_IGNORE_DEPRECATED_USAGE_END
 }
 
-API int LiveModelRequestMultiSlotDecisionWithFlags(livemodel_context_t* context, const char* event_id, const char* context_json, unsigned int flags, reinforcement_learning::multi_slot_response* resp, reinforcement_learning::api_status* status)
+API int LiveModelRequestMultiSlotDecisionWithFlags(livemodel_context_t* context, const char * event_id, const char * context_json, unsigned int flags, reinforcement_learning::multi_slot_response_detailed* resp, reinforcement_learning::api_status* status)
```
Comment (olgavrou) [inline bindings/cs/rl.net.native/rl.net.live_model.cc:139]: this should not be changed here right? this is still the request for multi slot, not multi slot detailed which was added below

File: bindings/cs/rl.net.native/rl.net.live_model.h [hunk 13]
```diff
@@ -26,33 +26,36 @@ typedef struct livemodel_context {
 // Global exports
 extern "C" {
     // NOTE: THIS IS NOT POLYMORPHISM SAFE!
```
Comment (olgavrou) [inline bindings/cs/rl.net.native/rl.net.live_model.h:28]: please revert the whitespace changes here

File: bindings/cs/rl.net/SlotRanking.cs [hunk 14]
```diff
@@ -0,0 +1,203 @@
...
+
+            public void Reset()
+            {
+                throw new NotSupportedException();
+            }
+        }
+    }
+}
+
+//        API slot_enumerator_adapter* CreateSlotEnumeratorAdapter(reinforcement_learning::slot_ranking* slot);
```
Comment (olgavrou) [inline bindings/cs/rl.net/SlotRanking.cs:196]: remove commented out code

File: include/slot_ranking.h [hunk 15]
```diff
@@ -0,0 +1,153 @@
+/**
+ * @brief slot_ranking definition. slot_ranking contains the cosen action, along with the action probability distribution for a slot.
```
Comment (olgavrou) [inline include/slot_ranking.h:2]: super super nit
```suggestion
 * @brief slot_ranking definition. slot_ranking contains the chosen action, along with the action probability distribution for a slot. 
```

File: include/slot_ranking.h [hunk 16]
```diff
@@ -0,0 +1,153 @@
...
+			*
+			* @param join_id The unique identifier for this interaction.  This event_id must be used when reporting the outcome for this action
+			*/
+		slot_ranking(char const* join_id);
+
+		/**
+		* @brief Unique join_id for this slot detailed instance.
+		* This join_id must be used when calling report_outcome so it can be joined with the chosen action
+		* @return const char*
+		*/
```
Comment (olgavrou) [inline include/slot_ranking.h:54]: join_id -> event_id? (same for the variable `_join_id`)
rlclientlib isn't really aware of joining afaik so it's a new term here.

File: rlclientlib/rlclientlib.vcxproj.filters [hunk 17]
```diff
@@ -136,7 +141,7 @@
     <None Include="schema\v2\CaEvent.fbs" />
     <None Include="schema\v2\MultiSlotEvent.fbs" />
     <None Include="schema\v2\Event.fbs" />
-    <None Include="schema\v2\DedupInfo.fbs" />
+    <None Include="ClassDiagram.cd" />
```
Comment (olgavrou) [inline rlclientlib/rlclientlib.vcxproj.filters:144]: looks like `DedupInfo.fbs` got overridden here

File: unit_test/live_model_test.cc [hunk 18]
```diff
@@ -796,11 +866,42 @@ BOOST_AUTO_TEST_CASE(ccb_explore_only_mode) {
...
+		BOOST_CHECK_EQUAL(action_id, 0);
+	}
+}
+
 const auto JSON_SLATES_CONTEXT = R"({"GUser":{"id":"a","major":"eng","hobby":"hiking"},"_multi":[{"TAction":{"a1":"f1"},"_slot_id":0},{"TAction":{"a2":"f2"},"_slot_id":0},{"TAction":{"a3":"f3"},"_slot_id":1},{"TAction":{"a4":"f4"},"_slot_id":1},{"TAction":{"a5":"f5"},"_slot_id":1}],"_slots":[{"Slot":{"a1":"f1"}},{"Slot":{"a2":"f2"}}]})";
 
 BOOST_AUTO_TEST_CASE(slates_explore_only_mode) {
   u::configuration config;
   cfg::create_from_json(JSON_CFG, config);
+  config.set(r::name::PROTOCOL_VERSION, "2");
```
Comment (olgavrou) [inline unit_test/live_model_test.cc:904]: does this mean that we don't test slates for v1 anymore?

File: unit_test/unit_test.vcxproj.filters [hunk 19]
```diff
@@ -24,9 +24,6 @@
     <ClInclude Include="common_test_utils.h">
       <Filter>Header Files</Filter>
     </ClInclude>
-    <ClInclude Include="mock_http_client.h">
```
Comment (olgavrou) [inline unit_test/unit_test.vcxproj.filters:27]: why was this deleted?

File: include/slot_ranking.h (hunk 16 above)
Comment (ormichae) [reply include/slot_ranking.h:54]: For ccb, all slots share an eventId, so the _slot_id uniquely identifies each slot for joining. The reason I introduced _join_id here is that slot_ranking is used in ranking_response, so I wanted to find a name that is applicable for both CB and CCB cases

File: unit_test/live_model_test.cc (hunk 18 above)
Comment (ormichae) [reply unit_test/live_model_test.cc:904]: That was added by mistake. Thanks for catching that

File: bindings/cs/rl.net.native/rl.net.live_model.cc (hunk 12 above)
Comment (ormichae) [reply bindings/cs/rl.net.native/rl.net.live_model.cc:139]: must have changed that by mistake, reverted now

File: rlclientlib/detailed_multi_slot_response.cpp [hunk 20]
```diff
@@ -0,0 +1 @@
+#include "detailed_multi_slot_response.h"
```
Comment (olgavrou) [inline rlclientlib/detailed_multi_slot_response.cpp:1]: looks like this file and the `.h` ones should be deleted?

File: rlclientlib/ClassDiagram.cd [hunk 21]
```diff
@@ -0,0 +1,2 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (olgavrou) [inline rlclientlib/ClassDiagram.cd:1]: remove file?

**Review by olgavrou — APPROVED** [review]
remove `unit_test/interaction.txt` and `unit_test/observation.txt`

File: rlclientlib/live_model_impl.cc [hunk 22]
```diff
@@ -243,6 +243,61 @@ namespace reinforcement_learning {
...
 
+  int live_model_impl::request_multi_slot_decision(const char * context_json, unsigned int flags, multi_slot_response_detailed& resp, api_status* status)
+  {
+	  const auto uuid = boost::uuids::to_string(boost::uuids::random_generator()());
+	  return request_multi_slot_decision(uuid.c_str(), context_json, flags, resp, status);
+  }
+
+  int live_model_impl::request_multi_slot_decision(const char * event_id, const char * context_json, unsigned int flags, multi_slot_response_detailed& resp, api_status* status)
+  {
+	  if (_learning_mode == APPRENTICE || _learning_mode == LOGGINGONLY) {
```
Comment (ataymano) [inline rlclientlib/live_model_impl.cc:254]: Looks like lines 254-290 are copy-paste from 201-236. Can we move it to some private method?

File: rlclientlib/live_model_impl.cc [hunk 23]
```diff
@@ -243,6 +247,36 @@ namespace reinforcement_learning {
...
+  int live_model_impl::request_multi_slot_decision(const char * event_id, const char * context_json, unsigned int flags, multi_slot_response_detailed& resp, api_status* status)
+  {
+    resp.clear();
+
+    std::vector<std::string> slot_ids;
+    std::vector<std::vector<uint32_t>> action_ids;
+    std::vector<std::vector<float>> action_pdfs;
+    std::string model_version;
+
+    RETURN_IF_FAIL(live_model_impl::request_multi_slot_decision_shared(event_id, context_json, slot_ids, action_ids, action_pdfs, model_version, status));
```
Comment (ataymano) [inline rlclientlib/live_model_impl.cc:265]: I suggest to rename it to maybe request_multi_slot_decision_impl instead of request_multi_slot_decision_shared.
Reasons:
- there is nothing inside this function about sharing something with someone, and we do not want to rename it if we suddenly decide to remove one of request_multi_slot_decision calls (and it stopped being shared actually)
- in this context it is a bit ambiguous with "shared" features (it has nothing to do with that, but can give that impression)

File: rlclientlib/live_model_impl.cc [hunk 24]
```diff
@@ -243,6 +247,36 @@ namespace reinforcement_learning {
...
+
+	  //set the size of buffer in response to match the number of slots
+	  resp.resize(slot_ids.size());
```
Comment (ataymano) [inline rlclientlib/live_model_impl.cc:268]: something is wrong with spaces here.

**Review by ataymano — APPROVED** [review]
